### PR TITLE
Tock 2.0 IPC

### PR DESCRIPTION
**Pull request by alevy**
### Pull Request Overview
This pull request ports the IPC driver to the Tock 2.0 driver interface.

This is a mostly fairly straightforward port. However, the old driver API relied on `allow` to return a discovered IPC endpoint upon success, which is no longer possible in the 2.0 driver interface (allow cannot return a value).

Recall that IPC works as follows:

1. A client "discovers" an IPC service by supplying the kernel with the service's name (taken from the process's TBF name header)
2. If such a service exists, the kernel supplies the client process with a 32-bit non-zero descriptor to reference the service (implemented as the AppId + 1, but that's not guaranteed).
3. The client can use this descriptor to share memory with the service (using allow) and to `notify` it (using command).

For discovery, `allow(IPC, 0)` had been used to both share a buffer containing the service name _and_ request that the kernel perform discovery, returning a descriptor on success.

_This is no longer possible_, and it also not canonical driver semantics.

With this PR, discovery is split into an `allow_readonly` (which is nice anyway so
package names can be hardcoded in ROM rather than copied to
memory) followed by a `command(IPC, 1)` to actually perform the discovery and return a descriptor.

To make this cleaner, I also rejiggerd the arguments in `command`, which now has to mean more than _just_ notify.

Previously, the arguments were:

1. Descriptor (as opposed to the normal command number)
2. Service-or-client-notify (which is actually kind of like a command number)
3. Unused

Now, the arguments are:

1. Command number (can be "discover," "service notify," or "client notify")
2. Target descriptor
3. Unused

There are a number of other changes I believe _could_ make sense to make to IPC before releasing 2.0, but I think they should be decoupled from this pull request, which merely adapts the interface very little to get it working. These are just ideas, and we should discuss all these _separately_:

- Allow both read-only and read-write buffer sharing between client/service
- Use borrow semantics for memory ownership
- Let client subscribe to be notified when a service is available
- Probably more...

### Testing Strategy

Tested using the IPC tutorial in libtock-c (examples/tutorial/05-ipc) after applying changes from tock/libtock-c#176

### TODO or Help Wanted

~~Still need to adapt libtock-c userspace to properly test.~~ (Done in tock/libtock-c#176)

Still need to update documentation

Please review not only the implementation but the change in interface described above. As I said, there are a variety of changes that might make sense for IPC that go beyond these minimal changes. I ask that we get something in that reliably reproduces the existing high-level semantics for IPC and after (or maybe concurrently) decide whether other changes should happen (either before or after 2.0).

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

**Comment by alevy** [comment]
Please do not merge yet, I have not tested properly and have not updated all the documentation. However, I believe it's more or less complete, so worth reviewing.

I've separated into two commits, one that changes the IPC driver and one (with a one-line change in a ton of files) that updates references to IPC in all boards using it. I suggest looking primarily at the first commit, which isn't actually that big.

**Comment by hudson-ayers** [comment]
This increases code size by enough that the kernel no longer fits into flash on Imix (hence the CI failure). Looking at hail, it increased size by 1264 bytes.

**Comment by alevy** [comment]
> 1264 bytes

I'll look into it, but that's _got_ to be an alignment thing (e.g., it maybe increased by enough of a margin to just push it over the alignment boundary or something), right?

File: kernel/src/ipc.rs
Comment (hudson-ayers) [inline kernel/src/ipc.rs:215]:
```suggestion
            0 =>
            /* Discover */
            {
```

File: kernel/src/ipc.rs
Comment (hudson-ayers) [inline kernel/src/ipc.rs:229]: Based on your description in this PR, and how this used to work when it was part of `allow()`, shouldn't this be `p.appid().id() + 1`?

File: kernel/src/ipc.rs
Comment (hudson-ayers) [inline kernel/src/ipc.rs:290]:
```suggestion

    /// If allow_readonly is called with target_id == 0, it is an IPC service discover
    /// call. The contents of the slice should be the string name of the IPC
    /// service. The IPC driver will store this name for use by the service discover
    /// command.
    fn allow_readonly(
```

File: kernel/src/ipc.rs
Comment (hudson-ayers) [inline kernel/src/ipc.rs:325]: I can't comment on the lines above this, but the doc comments are now out of date.

File: kernel/src/ipc.rs
Comment (hudson-ayers) [inline kernel/src/ipc.rs:328]: I am guessing this is the original reason that the target id was app_id + 1? Now that allow does not use target_id == 0, it seems we could go back to target_id = app_id. But maybe there is a benefit in keeping them separate in that it forces apps not to assume they are equal, which is something the kernel does not guarantee

**Review by hudson-ayers — COMMENTED** [review]
Looks mostly good to me, I have a few comments.

**Comment by hudson-ayers** [comment]
> > 1264 bytes
> 
> I'll look into it, but that's _got_ to be an alignment thing (e.g., it maybe increased by enough of a margin to just push it over the alignment boundary or something), right?

Using Phil's size tool, the code size required for functions in the `kernel::ipc::IPC::* ` namespace went from using 1006 bytes before this PR to using 2392 bytes afterwards. I think the size tool counts code space wasted due to alignment separately. Unfortunately the tool does not provide any resolution on where the size is going within individual methods on the struct. The one place in the changes where it is clear size might have changed is `command()` -- before, the code for service notify and client notify was almost entirely shared, and now it is duplicated across the two match arms. Still it makes no sense how that could over double the code size used..

File: kernel/src/ipc.rs
Comment (phil-levis) [inline kernel/src/ipc.rs:153]: 0 has to be the "exists" command; you should bump all the command IDs by 1:

https://github.com/tock/tock/blob/tock-2.0-dev/doc/reference/trd-syscalls.md#431-command-identifier-0

File: kernel/src/ipc.rs
Comment (phil-levis) [inline kernel/src/ipc.rs:214]: 0 has to be the "exists" command: you should bump all commands by 1:

https://github.com/tock/tock/blob/tock-2.0-dev/doc/reference/trd-syscalls.md#431-command-identifier-0

**Comment by phil-levis** [comment]
>     * Use borrow semantics for memory ownership

This is interesting; fundamentally the idea of this shared memory goes against Rust's ownership model, so does need to be protected by some sort of lock. If we do decide to follow the borrow semantics, though, this becomes message passing (except if it's read-only on both sides)?

File: kernel/src/ipc.rs
Comment (alevy) [reply kernel/src/ipc.rs:229]: Yes! Missed that thanks

**Comment by hudson-ayers** [comment]
> This is interesting; fundamentally the idea of this shared memory goes against Rust's ownership model, so does need to be protected by some sort of lock. If we do decide to follow the borrow semantics, though, this becomes message passing (except if it's read-only on both sides)?

I suspect that the way we use shared memory now would be difficult to use without undefined behavior for a rust userspace? I realize libtock-rs does not currently support IPC.

I am not sure what exactly defines "message passing", but it seems that the ability to 

- take a lock
- modify 1 byte of the shared memory 
- release the lock

is lower overhead than having to pass the entire shared memory region back as a message. Maybe this is more of a semantic difference

File: kernel/src/ipc.rs
Comment (alevy) [reply kernel/src/ipc.rs:58]: Unfortunately, the problem is not that it's not const, it's that `ReadWriteAppSlice` is not marked `Copy`, so without `const_in_array_repeat_expressions` (which is now not even an option with unstable) I don't think there are any tricks we can do.

If we're happy to make `ReadWriteAppSlice` `Copy`, we don't need unsafe, but I believe that could result in mutable aliases.

**Comment by phil-levis** [comment]
> This increases code size by enough that the kernel no longer fits into flash on Imix (hence the CI failure). Looking at hail, it increased size by 1264 bytes.

I've been banging up on this a bit with imix. We should do a spring cleaning and see where we could reclaim some space. ADC for example is 6.1k!

File: kernel/src/ipc.rs
Comment (hudson-ayers) [reply kernel/src/ipc.rs:58]: `const_in_array_repeat_expressions` was removed in favor of requiring explicit promotion of constants in constant blocks (see the discussion around [here](https://github.com/rust-lang/rust/issues/49147#issuecomment-736441091)). So as long as there is a const constructor, and you declare the const outside of the actual array instantiation, this works fine. This is the same reason the previous implementation works. Eventually this will get a little cleaner once [inline_const](https://github.com/rust-lang/rust/issues/76001) is stable -- then the `const` could be declared at the point of instantiation.

I tested the code changes above and they compile. Happy to push the commit if you prefer it, or drop it if you don't

**Comment by alevy** [comment]
@phil-levis:

> I've been banging up on this a bit with imix. We should do a spring cleaning and see where we could reclaim some space. ADC for example is 6.1k!

I totally agree, but also I'm not sure how reliable those numbers are at that granularity (the tool is great! don't get me wrong). For example, removing the nrf51 capsule from Hail _adds_ ~1kB to the kernel, and the scheduler module occupies ~10kB on Hail, but only ~5kB on the Nano BLE, and ~8kB on the nrf52840DKs

The code changes really _shouldn't_ require adding much code space at all, so I'll debug this enough to figure out what's going on (though again, removing things sometimes _increases_ the size, so it's pretty brittle) and get imix inline.

But agree that we should do a deeper more comprehensive dive into this. These modules are all far bigger than they need to be.

**Comment by phil-levis** [comment]
> the scheduler module occupies ~10kB on Hail, but only ~5kB on the Nano BLE, and ~8kB on the nrf52840DKs

I think this is inlining: modules higher in the call tree can take on code from lower ones. Happens less in the opposite direction.

File: kernel/src/ipc.rs
Comment (bradjc) [inline kernel/src/ipc.rs:114]: We should probably look at https://github.com/tock/tock/issues/1993 since we are here. What happens if `add_mpu_region` fails?

File: kernel/src/ipc.rs
Comment (alevy) [reply kernel/src/ipc.rs:58]: Aha! I didn't realize the subtlety about declaring the const ouside the array initialization so when I tried it it didn't work. This seems much much much better than what I have, will change.

Comment (alevy) [reply kernel/src/ipc.rs:58]: Coincidentally also fixes the imix code size issue

File: kernel/src/ipc.rs
Comment (alevy) [reply kernel/src/ipc.rs:328]: Right, I hadn't thought through all the implications so just left it as is for now. After porting libtock-c, I'll see if we can just elide this transformation. I don't think there is a significant benefit to obscuring this intentionally from apps if it doesn't serve another purpose (the semantics should still not promise anything about the meaning of the numbers though)

**Review by hudson-ayers — APPROVED** [review]

File: kernel/src/sched.rs
Comment (hudson-ayers) [reply kernel/src/sched.rs:328]: At some point in the force pushing this change seems to have been dropped

**Comment by alevy** [comment]
Please still don't merge, I'm currently testing with libtock-c and still need to update documentation

**Comment by bradjc** [comment]
I was playing around with IPC today and I found some issues. https://github.com/tock/tock/issues/1993 Might not be relevant for this PR but wanted to make a note of it.

**Comment by alevy** [comment]
I pushed a commit (which I will squash before we merge) that fixes a bug I had introduced in `schedule_callback`. Because the bug was subtle and hard to catch, I also updated the names of the arguments. While after testing I'm fairly certain that function is now correct, it would be very useful for someone to quickly review my changes (just in that last commit should be fine) and verify that what's going on in that function kind of actually makes sense.

**Comment by alevy** [comment]
Testing complete. Just going to update documentation, then I think we should be good (modulo squashing the last commit with the first one)

**Comment by alevy** [comment]
Documentation updated

**Review by hudson-ayers — APPROVED** [review]
The most recent commit looks correct to me, it seems to match the behavior from when `expose_to()` was used, and the updated names are helpful

**Comment by alevy** [comment]
Squashed

File: kernel/src/mem.rs
Comment (bradjc) [inline kernel/src/mem.rs:135]: Can you capture some wisdom on what this is?

File: kernel/src/sched.rs
Comment (bradjc) [inline kernel/src/sched.rs:686]: Are we adding this to #1993?

Comment (alevy) [reply kernel/src/sched.rs:686]: Yes

**Review by bradjc — APPROVED** [review]

**Comment by phil-levis** [comment]
Hurray!

**Pull request merged**
